### PR TITLE
KAA-1225: [Objective-C SDK] Incorrect calculation of log delivery time

### PR DESCRIPTION
**Pull request by hamsternik**
There were several important changes carried out while performing this ticket (except fixing incorrect calculation of log delivery time and aditing test case):
- changed implementation of notifyOnSuccessDeliveryRunnersWithBucketInfo method according to C++ KAA SDK [method implementation](https://github.com/kaaproject/kaa/blob/1d429a30bb4b5206376b740bb21483929a881ace/client/client-multi/client-cpp/impl/log/LogCollector.cpp#L410) and [its next usage](https://github.com/kaaproject/kaa/blob/develop/client/client-multi/client-cpp/impl/log/LogCollector.cpp#L325).
- added public getter for deliveryRunnerDictionary from AbstractLogCollector class
- added receivedResponseTime property to return the total delivery time of log (in milliseconds).
- implemented new category for NSDate class
- made small code refactoring

### CODE REVIEW

**Comment by hamsternik** [comment]
@abohomol @am1kk please review

File: client/client-multi/client-objective-c/Kaa/utils/NSDate+Timestamp.m [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#import "NSDate+Timestamp.h"
+#import <QuartzCore/QuartzCore.h>
+
+@implementation NSDate (Timestamp)
+
++ (double)currentTimeInMilliseconds {
+    return CACurrentMediaTime() * 1000;
+}
+
+@end
```
Comment (am1kk) [inline client/client-multi/client-objective-c/Kaa/utils/NSDate+Timestamp.m:26]: File must ends with new line.

File: client/client-multi/client-objective-c/KaaTests/logging/DefaultLogUploadStrategyTest.m [hunk 2]
```diff
@@ -221,5 +228,4 @@ - (void)maxParallelUploadCountInDecisionHelper:(int64_t)maxParallelUploads {
     [verifyCount(logTransport, times(maxParallelUploads)) sync];
 }
 
-
-@end
+@end
```
Comment (am1kk) [inline client/client-multi/client-objective-c/KaaTests/logging/DefaultLogUploadStrategyTest.m]: File must ends with new line.

File: client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	    }
+    }];
+    [NSThread sleepForTimeInterval:0.1];
+    
+    for (NSInteger i = 0; i < kIterCount; i++) {
+        XCTAssertLessThan(bucketInfoInstances[i].bucketDeliveryDuration, 1.f);
+    }
+}
+
+@end
```
Comment (am1kk) [inline client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m]: File must ends with new line.

**Review by am1kk — CHANGES_REQUESTED** [review]

File: client/client-multi/client-objective-c/Kaa/utils/NSDate+Timestamp.m (hunk 1 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/Kaa/utils/NSDate+Timestamp.m:26]: Done.

File: client/client-multi/client-objective-c/KaaTests/logging/DefaultLogUploadStrategyTest.m (hunk 2 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/KaaTests/logging/DefaultLogUploadStrategyTest.m]: Done.

File: client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m (hunk 3 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m]: Done.

**Review by am1kk — APPROVED** [review]

File: client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m [hunk 4]
```diff
@@ -138,29 +144,31 @@ - (void)onLogResponse:(LogSyncResponse *)response {
...
                     [self.bucketInfoDictionary removeObjectForKey:key];
 
                     if (status.result == SYNC_RESPONSE_RESULT_TYPE_SUCCESS) {
+	                    bucketInfo.receivedResponseTime = [NSDate currentTimeInMilliseconds];
+                        
                         [self.storage removeBucketWithId:status.requestId];
                         
                         [[self.executorContext getCallbackExecutor] addOperationWithBlock:^{
                             [weakSelf notifyOnSuccessDeliveryRunnersWithBucketInfo:bucketInfo];
+				            [self.deliveryRunnerDictionary removeObjectForKey:@(bucketInfo.bucketId)];
```
Comment (abohomol) [inline client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m]: Incorrect indention

File: client/client-multi/client-objective-c/Kaa/log/BucketRunner.m [hunk 5]
```diff
@@ -14,11 +14,13 @@
  * limitations under the License.
  */
 
+#import <QuartzCore/QuartzCore.h>
```
Comment (abohomol) [inline client/client-multi/client-objective-c/Kaa/log/BucketRunner.m]: Do we still need this import after we incapsulated time calculation inside NSDate category?

File: client/client-multi/client-objective-c/KaaTests/logging/LogDeliveryCallbackTests.m [hunk 6]
```diff
@@ -66,20 +66,23 @@ @implementation LogDeliveryCallbackTests
...
                                                         executorContext:self.executorContext
                                                          channelManager:self.channelManager
                                                         failoverManager:self.failoverManager];
     
-    self.strategy = mockProtocol(@protocol(LogUploadStrategy));
     [given([self.strategy getMaxParallelUploads]) willReturn:@(10)];
     [self.logCollector setValue:self.strategy forKey:@"strategy"];
 }
 
+// TODO: NK: Separate this testcase for several sub-testcases (focus on status result)
```
Comment (abohomol) [inline client/client-multi/client-objective-c/KaaTests/logging/LogDeliveryCallbackTests.m]: Possibly redundant/outdated comment

File: client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m [hunk 7]
```diff
@@ -22,6 +22,8 @@
 #import "FailoverManager.h"
 #import "KaaLogging.h"
 #import "LogCollector.h"
+#import <QuartzCore/QuartzCore.h>
```
Comment (abohomol) [inline client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m]: Do we still need this import after we incapsulated time calculation inside NSDate category?

File: client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.h [hunk 8]
```diff
@@ -39,8 +39,10 @@
 
 - (void)uploadIfNeeded;
 
-- (void)addDeliveryRunner:(BucketRunner *)runner bucketInfo:(BucketInfo *)bucketInfo;
+- (void)addDeliveryRunner:(BucketRunner *)runner byBucketInfoKey:(NSNumber *)bucketInfoKey;
 
 - (void)notifyOnSuccessDeliveryRunnersWithBucketInfo:(BucketInfo *)bucketInfo;
 
+- (NSMutableDictionary *)getDeliveryRunnerDictionary;
```
Comment (abohomol) [inline client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.h]: I'm not seeing that method to be used anywhere outside the AbstractLogCollector class.

File: client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m (hunk 4 above)
Comment (abohomol) [inline client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m]: As far as I see this method is just moved here out of notifyOnSuccessDeliveryRunnersWithBucketInfo: method.
Is there any reason for that?

File: client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    
+    LogDeliveryStatus *status = [[LogDeliveryStatus alloc] init];
+    status.result = SYNC_RESPONSE_RESULT_TYPE_SUCCESS;
+    status.requestId = 42;
+    
+    const NSInteger kIterCount = 10;
+    
+    NSMutableArray<BucketInfo *> *bucketInfoInstances = [NSMutableArray new];
+    for (int i = 0; i < kIterCount; i++) {
+	    BucketInfo *bucketInfo = [[BucketInfo alloc] initWithBucketId:status.requestId logCount:1];
```
Comment (abohomol) [inline client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m]: Incorrect indention in this line and the one below

File: client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+	    bucketInfo.receivedResponseTime = [NSDate currentTimeInMilliseconds];
+        [bucketInfoInstances addObject:bucketInfo];
+    }
+    
+    logCollector.bucketInfoDictionary[@(status.requestId)] = bucketInfoInstances;
+    
+    for (NSInteger i = 0; i < kIterCount; i++) {
+	    BucketRunner *brunner = [[BucketRunner alloc] init];
```
Comment (abohomol) [inline client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m]: Incorrect indention

File: client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+	    [logCollector addDeliveryRunner:brunner byBucketInfoKey:@([[bucketInfoInstances objectAtIndex:i] bucketId])];
+    }
+    
+    [[executorContext getCallbackExecutor] addOperationWithBlock:^{
+	    for (NSInteger i = 0; i < kIterCount; i++) {
```
Comment (abohomol) [inline client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m]: Incorrect indention

**Review by abohomol — CHANGES_REQUESTED** [review]
There are several things that aren't obvious for me. I'd ask for some clarifications on those and insist on fixing minor code style issue.

File: client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m (hunk 7 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m]: Of course, no. Removed redundant import.

File: client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m (hunk 4 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m]: Fixed.

Comment (hamsternik) [reply client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.m]: Firstly, the goal of the _notifyOnSuccessDeliveryRunnersWithBucketInfo_ method to setup passed bucket info instance to each of runners stored by concrete bucket info ID.
The next reason is unreasonable removing all runners from local storage (self.deliveryRunnerDictionary). 

So, without moving this line out from _notifyOnSuccessDeliveryRunnersWithBucketInfo_ method, I can't store my bucket info's instances for some more time. And as you can see in **LogDeliveryTest** single test case, I store all bucketInfos for next checking of their bucketDeliveryDuration.

File: client/client-multi/client-objective-c/Kaa/log/BucketRunner.m (hunk 5 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/Kaa/log/BucketRunner.m]: This import is redundant too. Removed.

File: client/client-multi/client-objective-c/KaaTests/logging/LogDeliveryCallbackTests.m (hunk 6 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/KaaTests/logging/LogDeliveryCallbackTests.m]: Waste comment. Removed.

File: client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.h (hunk 8 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/Kaa/log/AbstractLogCollector.h]: Yes, this method had been used at debug process and hasn't been removed after bug fixing. Removed

File: client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m (hunk 9 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m]: Fixed.

File: client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m (hunk 10 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m]: Fixed.

File: client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m (hunk 11 above)
Comment (hamsternik) [reply client/client-multi/client-objective-c/KaaTests/logging/delivery/LogDeliveryTest.m]: Fixed.

**Comment by abohomol** [comment]
LGTM

**Comment by rasendubi** [comment]
@kaa-leeroy retest this please

**Pull request merged**